designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,300 @@
+Alternatively, this could be handled in a way that [helm kustomize plugin](https://helm.sh/docs/topics/advanced/#post-rendering) discovers the plugin based on the non-existence of a separator in the path provided, in which case Kubebuilder will search in `$PATH`, otherwise resolve any relative paths to a fully qualified path.
+
+* Pros
+  * This provides flexibility for the user to specify the file paths that the plugin would be placed in and Kubebuilder could discover the binaries in those user specified file paths.
+
+  * No constraints on plugin binary naming or directory placements from the Kubebuilder side.
+
+  * Provides a default value for the plugin directory in case user wants to use that to drop their plugins.
+
+#### Approach #3
+
+Another approach I would like to mention is adding plugin executables with a prefix `kubebuilder-` followed by the plugin name to the PATH variable. This will enable KB to traverse through the PATH looking for the plugin executables starting with the prefix `kubebuilder-` and matching by the plugin name that was provided in the CLI. Furthermore, a check should be added to verify that the match is an executable or not and return an error if it's not an executable. This approach provides a lot of flexibility in terms of plugin discovery as all the user needs to do is to add the plugin executable to the PATH and Kubebuilder will discover it.
+
+* Pros
+  * `kubectl` and `git` follow the same approach for discovering plugins, so there’s prior art.
+
+  * There’s a lot of flexibility in just dropping plugin binaries to PATH variable and enabling the discovery without having to enforce any other constraints on the placements of the plugins.
+
+* Cons
+  * Enumerating the list of all available plugins might be a bit tough compared to having a single folder with the list of available plugins and having to enumerate those.
+
+  * These plugin binaries cannot be run in a standalone manner outside of Kubebuilder, so may not be very ideal to add them to the PATH var.
+
+#### Open questions
+
+* Do we want to support the addition of new arbitrary subcommands other than the subcommands (init, create api, create webhook) that we already support?
+  * Not for the EP or initial implementation, but can revisit later.
+
+* Do we need to discover flags by calling the plugin binary or should we have users define them in the project configuration?
+  * Flags will have to be discovered by calling the plugin binary.
+
+* What alternatives to stdin/stdout exist and why shouldn't we use them?
+  * Other alternatives exist such as named pipe and sockets, but stdin/stdout seems to be more suitable for our needs.
+
+* What happens when two plugins bind the same flag name? Will there be any conflicts?
+
+* Should plugins depend on environment variables?
+  * The same plugin would behave differently based on the environment.
+
+* Should the API version be a part of the plugin request spec?
```

Comment:
Which API version? For `PluginRequest` and `PluginResponse`? It would be nice to encode these actually, so we can mark the initial schemas as alpha (`v1alpha1`).